### PR TITLE
Gas Price Algorithm

### PR DESCRIPTION
**Pull request by MitchTurner**
Closes: https://github.com/FuelLabs/fuel-core/issues/1957

Part of https://github.com/FuelLabs/fuel-core/issues/1624

Add the gas price algorithm with some tools for analyzing its behavior.

The algorithm is made up of two portions:
- `AlgorithmUpdaterV1`
- `AlgorithmV1`

The updater holds the state of the historical data and can generate new algorithms as its data is updated. The algorithm just takes the `block_bytes` and can calculate the price from that data.

I've included a lot of docs on the actual code, so see that for more details. LMK if something is missing :)

## Checklist
- [ ] Breaking changes are clearly marked as such in the PR description and changelog
- [x] New behavior is reflected in tests
- [ ] [The specification](https://github.com/FuelLabs/fuel-specs/) matches the implemented behavior (link update PR if changes are needed)

### Before requesting review
- [x] I have reviewed the code myself
- [ ] I have created follow-up issues caused by this PR and linked them here

### After merging, notify other teams

[Add or remove entries as needed]

- [ ] [Rust SDK](https://github.com/FuelLabs/fuels-rs/)
- [ ] [Sway compiler](https://github.com/FuelLabs/sway/)
- [ ] [Platform documentation](https://github.com/FuelLabs/devrel-requests/issues/new?assignees=&labels=new+request&projects=&template=NEW-REQUEST.yml&title=%5BRequest%5D%3A+) (for out-of-organization contributors, the person merging the PR will do this)
- [ ] Someone else?

### CODE REVIEW

**Comment by MitchTurner** [comment]
It sounds like we aren't planning on having any DA costs in the immediate future. So I think we put a pin in the research for the DA projection stuff and move forward. I could spend weeks on improving DA gas price portion. The Exec Gas Price portion is much simpler and all we need for now.

If we set the costs to 0 in prod, then this algo will work just fine :)

From here we can get the updater and algo installed into our code and we can make adjustments as they come up. The general shape is correct.

**Comment by MitchTurner** [comment]
Need to disallow unchecked arithmetic.

File: crates/gas-price-algorithm/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
+use plotters::prelude::*;
```
Comment (xgreenx) [inline crates/gas-price-algorithm/src/main.rs:1]: Maybe we could move it to own binary

**Comment by MitchTurner** [comment]
Looks like the checked arithmetic broke some things. I'm debugging.

File: crates/gas-price-algorithm/gas-price-analysis/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "gas-price-analysis"
+version = "0.1.0"
+edition = "2021"
+
+[workspace]
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
```
Comment (Dentosal) [inline crates/gas-price-algorithm/gas-price-analysis/Cargo.toml:9]:
```suggestion
```

File: crates/gas-price-algorithm/gas-price-analysis/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "gas-price-analysis"
+version = "0.1.0"
+edition = "2021"
+
```
Comment (Dentosal) [inline crates/gas-price-algorithm/gas-price-analysis/Cargo.toml:5]:
```suggestion
publish = false

```

File: crates/gas-price-algorithm/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+[package]
+name = "gas-price-algorithm"
```
Comment (Dentosal) [inline crates/gas-price-algorithm/Cargo.toml:2]:
```suggestion
name = "fuel-gas-price-algorithm"
```

File: Cargo.toml [hunk 5]
```diff
@@ -81,6 +83,7 @@ fuel-core-tests = { version = "0.0.0", path = "./tests" }
 fuel-core-upgradable-executor = { version = "0.30.0", path = "./crates/services/upgradable-executor" }
 fuel-core-wasm-executor = { version = "0.30.0", path = "./crates/services/upgradable-executor/wasm-executor", default-features = false }
 fuel-core-xtask = { version = "0.0.0", path = "./xtask" }
+gas-price-algorithm = { version = "0.29.0", path = "./crates/gas-price-algorithm" }
```
Comment (Dentosal) [inline Cargo.toml:86]:
```suggestion
fuel-gas-price-algorithm = { version = "0.29.0", path = "./crates/gas-price-algorithm" }
```

**Comment by MitchTurner** [comment]
@Dentosal I made some small functional changes wrt minimum gas prices.

File: crates/fuel-gas-price-algorithm/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+    }
+}
+
+/// The state of the algorithm used to update the gas price algorithm for each block
+///
+/// Because there will always be a delay between blocks submitted to the L2 chain and the blocks
+/// being recorded on the DA chain, the updater needs to make "projections" about the cost of
+/// recording any given block to the DA chain. This is done by tracking the cost per byte of recording
+/// for the most recent blocks, and using the known bytes of the unrecorded blocks to estimate
+/// the cost for that block. Every time the DA recording is updated, that the projections are recalculated.
```
Comment (MitchTurner) [inline crates/fuel-gas-price-algorithm/src/lib.rs:139]:
```suggestion
/// the cost for that block. Every time the DA recording is updated, the projections are recalculated.
```

File: crates/fuel-gas-price-algorithm/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+///
+/// This projection will inevitably lead to error in the gas price calculation. Special care should be taken
+/// to account for the worst case scenario when calculating the parameters of the algorithm.
+pub struct AlgorithmUpdaterV1 {
+    /// The gas price for to cover the execution of the next block
```
Comment (MitchTurner) [inline crates/fuel-gas-price-algorithm/src/lib.rs:144]:
```suggestion
    /// The gas price to cover the execution of the next block
```

**Review by Dentosal — CHANGES_REQUESTED** [review]
Still some change suggestions unapplied

File: crates/gas-price-algorithm/gas-price-analysis/Cargo.toml (hunk 2 above)
Comment (MitchTurner) [reply crates/gas-price-algorithm/gas-price-analysis/Cargo.toml]: I just removed this.

**Review by Dentosal — APPROVED** [review]

**Review by Dentosal — APPROVED** [review]

**Pull request merged**